### PR TITLE
Pipeline: support fine grained window and window sort

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #6518 

Problem Summary:

### What is changed and how it works?

wait for https://github.com/pingcap/tiflash/pull/6934

- move some codes from `MergeSortingBlockInputStream` to `SortHelper`.
- rename `TopNTransformOp` to `LocalSortTransformOp` and refine it.
- replace `builder.appendTransformOp(std::make_unique<ExpressionTransformOp>(exec_status, log->identifier(), before_sort_actions));` by `InterpreterUtils::executeExpression`.
- add `executeLocalSort` to append `LocalSortTransformOp` or `LocalLimitTransformOp`(order by const).
- Add `WindowTransformOp` to support window function.
- Add some fine grained shuffle unit tests for window.
- add more unit tests for topN

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Operators/LocalSortTransformOp.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            }
+            return OperatorStatus::HAS_OUTPUT;
+        }
+        SortHelper::removeConstantsFromBlock(block);
+        sortBlock(block, order_desc, limit);
+        sorted_blocks.emplace_back(std::move(block));
+        return OperatorStatus::NEED_INPUT;
+    }
+    case LocalSortStatus::MERGE:
+        throw Exception("Unexpcet status: MERGE.");
```
Comment (SeaRise) [inline dbms/src/Operators/LocalSortTransformOp.cpp:70]:
```suggestion
        throw Exception("Unexpected status: MERGE.");
```

**Comment by SeaRise** [comment]
git merge https://github.com/pingcap/tiflash/pull/6934

File: dbms/src/Flash/Pipeline/Pipeline.cpp [hunk 2]
```diff
@@ -237,11 +237,9 @@ bool Pipeline::isSupported(const tipb::DAGRequest & dag_request)
             switch (executor.tp())
             {
             case tipb::ExecType::TypeTableScan:
-                if (executor.tbl_scan().keep_order())
-                {
-                    is_supported = false;
-                    return false;
-                }
+                // // TODO support keep order.
```
Comment (SeaRise) [inline dbms/src/Flash/Pipeline/Pipeline.cpp:240]:
```suggestion
                // TODO support keep order.
```

File: dbms/src/Operators/LocalSortTransformOp.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+OperatorStatus LocalSortTransformOp::transformImpl(Block & block)
+{
+    switch (status)
+    {
+    case LocalSortStatus::PARTIAL:
+    {
+        /// If there were only const columns in sort description, then there is no need to sort, just cut block.
+        if (order_desc.empty())
+        {
+            cut(block, limit);
```
Comment (windtalker) [inline dbms/src/Operators/LocalSortTransformOp.cpp:65]: Does this mean it will return all the input blocks after cut?

Comment (SeaRise) [reply dbms/src/Operators/LocalSortTransformOp.cpp:65]: yes, but here only a single block will be cutted, as with partial sort stream.
https://github.com/pingcap/tiflash/blob/e2f16fa997fbb551f5c709c1b455ff7e58ce5a03/dbms/src/DataStreams/MergeSortingBlockInputStream.cpp#L114-L117

File: dbms/src/Operators/LocalSortTransformOp.cpp [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+namespace DB
+{
+namespace
+{
+void cut(Block & block, size_t limit)
+{
+    if unlikely (!block)
+        return;
+    if (block.rows() <= limit)
```
Comment (SeaRise) [inline dbms/src/Operators/LocalSortTransformOp.cpp:28]:
```suggestion
    if (limit > 0 && block.rows() <= limit)
```

File: dbms/src/Operators/LocalSortTransformOp.cpp (hunk 3 above)
Comment (SeaRise) [reply dbms/src/Operators/LocalSortTransformOp.cpp]: add `executeLocalSort` to append `LocalSortTransformOp` or `LocalLimitTransformOp`(order by const).

File: dbms/src/DataStreams/SortHelper.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            --columns;
+        }
+        else
+            ++i;
+    }
+}
+
+void removeConstantsFromSortDescription(const Block & header, SortDescription & description)
+{
+    description.erase(
```
Comment (windtalker) [inline dbms/src/DataStreams/SortHelper.cpp:37]: Although not releated to your newly added code, but there maybe a potential problem that the const column in `header` and `block` is not the same, I think we need to record the const column in `header` as the meta data, then only remove the column recoreded when call `removeConstantsFromBlock`?

File: dbms/src/DataStreams/SortHelper.cpp [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        }),
+        description.end());
+}
+
+bool isSortByConst(const Block & header, const SortDescription & description)
+{
+    return std::all_of(description.begin(), description.end(), [&](const SortColumnDescription & elem) {
+        return elem.column_name.empty()
+            ? header.safeGetByPosition(elem.column_number).column->isColumnConst()
+            : header.getByName(elem.column_name).column->isColumnConst();
```
Comment (windtalker) [inline dbms/src/DataStreams/SortHelper.cpp:52]: Maybe the code style should be the same in `isSortByConst` and `removeConstantsFromSortDescription`

File: dbms/src/Flash/Coprocessor/InterpreterUtils.cpp [hunk 7]
```diff
@@ -162,6 +166,35 @@ void orderStreams(
...
+    const Context & context,
+    const LoggerPtr & log)
+{
+    auto input_header = group_builder.getCurrentHeader();
+    if (SortHelper::isSortByConst(input_header, order_descr))
+    {
+        // For order by const and limit != 0, we will generate LimitOperator directly.
+        if (limit != 0)
+        {
+            auto global_limit = std::make_shared<GlobalLimitTransformAction>(input_header, limit);
```
Comment (windtalker) [inline dbms/src/Flash/Coprocessor/InterpreterUtils.cpp:183]: Why use `global_limit` when the sort is local sort?

Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/InterpreterUtils.cpp]: ok, use `local_limit` instead

File: dbms/src/DataStreams/SortHelper.cpp (hunk 5 above)
Comment (SeaRise) [reply dbms/src/DataStreams/SortHelper.cpp:37]: I prefer to modify in another pr..

**Review by windtalker — APPROVED** [review]
LGTM

File: dbms/src/DataStreams/SortHelper.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace DB::SortHelper
+{
+/** Remove constant columns from block.
+  */
+void removeConstantsFromBlock(Block & block);
+
+void removeConstantsFromSortDescription(const Block & header, SortDescription & description);
+
+bool isSortByConst(const Block & header, const SortDescription & description);
```
Comment (gengliqi) [inline dbms/src/DataStreams/SortHelper.h:26]: Add comments for these functions?

File: dbms/src/Flash/Coprocessor/InterpreterUtils.cpp [hunk 9]
```diff
@@ -162,6 +166,35 @@ void orderStreams(
...
+    {
+        // For order by const and limit != 0, we will generate LimitOperator directly.
+        if (limit != 0)
+        {
+            auto local_limit = std::make_shared<LocalLimitTransformAction>(input_header, limit);
+            group_builder.transform([&](auto & builder) {
+                builder.appendTransformOp(std::make_unique<LimitTransformOp<LocalLimitPtr>>(exec_status, log->identifier(), local_limit));
+            });
+        }
+        // For order by const and limit == 0, do nothing here.
```
Comment (gengliqi) [inline dbms/src/Flash/Coprocessor/InterpreterUtils.cpp:188]: Should it return an empty block if limit is zero? If so, is it expected that do nothing here?

**Review by ywqzzy — APPROVED** [review]

File: dbms/src/Operators/LocalSortTransformOp.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    {
+    case LocalSortStatus::PARTIAL:
+    {
+        if unlikely (!block)
+        {
+            // convert to merge phase.
+            status = LocalSortStatus::MERGE;
+            if likely (!sorted_blocks.empty())
+            {
+                merge_impl = std::make_unique<MergeSortingBlocksBlockInputStream>(
```
Comment (gengliqi) [inline dbms/src/Operators/LocalSortTransformOp.cpp:52]: The name `MergeSortingBlocksBlockInputStream ` is confusing. 
How about `MergeSortingBlocksTransformAction`?

File: dbms/src/DataStreams/SortHelper.h (hunk 8 above)
Comment (SeaRise) [reply dbms/src/DataStreams/SortHelper.h]: ok, both added.

File: dbms/src/Operators/LocalSortTransformOp.cpp (hunk 10 above)
Comment (SeaRise) [reply dbms/src/Operators/LocalSortTransformOp.cpp:60]: The `MergeSortingBlocksBlockInputStream` is used in pull model to do merge sort.
I add a comment that `TODO refine MergeSortingBlocksBlockInputStream and use a more common class to do merge sort in both push model and pull model`.

**Review by gengliqi — APPROVED** [review]

**Pull request merged**